### PR TITLE
SALTO-4807 - Create references in strings inside article_body and actions

### PR DESCRIPTION
**Pull request by idoser96**
If there is a link inside article_body (without a brand), or inside action of trigger, automation and macro, check for regex and replace ids with reference expressions

---

this is by config
also need to update some docs about the new possible config?
requires noise reduction for the change from brand_url to brand reference expression

---
_Release Notes_: 
Zendesk Adapter:
* in elements, in string fields, it is now possible to change ids in links with reference expressions

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/zendesk-adapter/src/filters/article/article_body.ts [hunk 1]
```diff
@@ -16,45 +16,22 @@
...
 
 const log = logger(module)
 const { awu } = collections.asynciterable
-const { isDefined } = lowerDashValues
-const { createMissingInstance } = referencesUtils
-
 
 const BODY_FIELD = 'body'
-
-const ELEMENTS_REGEXES = [
```
Comment (idoser96) [inline packages/zendesk-adapter/src/filters/article/article_body.ts:46]: Moved to utils.ts

File: packages/zendesk-adapter/src/filters/article/article_body.ts [hunk 2]
```diff
@@ -168,16 +90,24 @@ const updateArticleTranslationBody = ({
...
-          urlPart,
-          urlBrandInstance,
-          additionalInstances,
-          enableMissingReferences,
-        })
+        urlPart => {
+          const urlSubdomain = urlPart.match(DOMAIN_REGEX)?.pop()
+          // We already made sure that the brand exists, so we can just return it
+          if (urlBrandInstance && urlSubdomain !== undefined) {
+            return [new ReferenceExpression(urlBrandInstance.elemID.createNestedID('brand_url'), urlBrandInstance?.value.brand_url)]
```
Comment (idoser96) [inline packages/zendesk-adapter/src/filters/article/article_body.ts:102]: This was extracted from transformReferenceUrls

File: packages/zendesk-adapter/src/config.ts [hunk 3]
```diff
@@ -2801,6 +2802,7 @@ export const configType = createMatchingObjectType<Partial<ZendeskConfig>>({
           appReferenceLocators: { refType: IdLocatorType },
           guide: { refType: GuideType },
           resolveOrganizationIDs: { refType: BuiltinTypes.BOOLEAN },
+          transformLinks: { refType: BuiltinTypes.BOOLEAN },
```
Comment (netama) [inline packages/zendesk-adapter/src/config.ts:2805]: nit: maybe `extractReferencesFromFreeText` or `extractFreeTextReferences`?

**Review by netama — COMMENTED** [review]
looks good, added some comments and questions

File: packages/zendesk-adapter/src/constants.ts [hunk 4]
```diff
@@ -78,7 +78,7 @@ export const TRIGGER_TYPE_NAME = 'trigger'
 export const TRIGGER_CATEGORY_TYPE_NAME = 'trigger_category'
 export const LOCALE_TYPE_NAME = 'locale'
 export const TICKET_STATUS_CUSTOM_STATUS_TYPE_NAME = 'custom_status'
-export const DEFLECTION_TYPE = 'deflection'
+export const DEFLECTION_NOTIFICATION = 'deflection'
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/constants.ts:81]: Because 'TYPE" sounds more like a Zendesk type, and i want to prevent that confusion

File: packages/zendesk-adapter/test/filters/article/article_body.test.ts [hunk 5]
```diff
@@ -200,18 +229,18 @@ describe('article body filter', () => {
...
-        const missingArticleInstance = createMissingInstance(ZENDESK, ARTICLE_TYPE_NAME, `${brandName}_124`)
-        const missingSectionInstance = createMissingInstance(ZENDESK, SECTION_TYPE_NAME, `${brandName}_123`)
-        const missingCategoryInstance = createMissingInstance(ZENDESK, CATEGORY_TYPE_NAME, `${brandName}_123`)
-        const missingArticleAttachmentInstance = createMissingInstance(ZENDESK, ARTICLE_ATTACHMENT_TYPE_NAME, `${brandName}_123`)
-        missingArticleInstance.value.id = '124'
-        missingSectionInstance.value.id = '123'
-        missingCategoryInstance.value.id = '123'
-        missingArticleAttachmentInstance.value.id = '123'
-        const fetchedTranslationWithoutReferences = elements.filter(isInstanceElement).find(i => i.elemID.name === 'translationWithEmptyBrand')
+        const missingArticleInstance = createMissingInstance(ZENDESK, ARTICLE_TYPE_NAME, `${brandName}_0`)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/test/filters/article/article_body.test.ts:235]: This checks for missings, because i randomize ids in the tests, i use '0' as a 'must be missing' id

File: packages/zendesk-adapter/src/filters/article/article_body.ts [hunk 6]
```diff
@@ -222,30 +152,25 @@ const getWarningsForMissingBrands = (
...
-    [ARTICLE_ATTACHMENTS_FIELD]:
-        _.keyBy(
-          instances.filter(e => e.elemID.typeName === ARTICLE_ATTACHMENT_TYPE_NAME && e.value.id !== undefined),
-          i => _.toString(i.value.id)
-        ),
-  }
-  const includedBrands = getBrandsForGuide(instances, config[FETCH_CONFIG])
+  const instancesById = _.keyBy(
+    instances.filter(instance => _.isNumber(instance.value.id)),
+    i => parseInt(i.value.id, 10)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/article/article_body.ts:157]: For some reason _keyBy has an issue that if the key is of unknown type, it thinks the values are InstanceElement | number
this parseInt is used for type check (its either that or write _.keyBy<InstanceElement>()

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts [hunk 7]
```diff
@@ -284,8 +312,16 @@ const getContainers = (instances: InstanceElement[]): TemplateContainer[] =>
...
-const replaceFormulasWithTemplates = (instances: InstanceElement[], enableMissingReferences?: boolean): void => {
+const replaceFormulasWithTemplates = (
+  instances: InstanceElement[],
+  enableMissingReferences?: boolean,
+  transformLinks?: boolean
+): void => {
   const instancesByType = _.groupBy(instances, i => i.elemID.typeName)
+  const instancesById = _.keyBy(
+    instances.filter(i => _.isNumber(i.value.id)),
+    i => _.parseInt(i.value.id)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/handle_template_expressions.ts:323]: talked IRL

File: packages/zendesk-adapter/src/filters/article/article_body.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/article/article_body.ts:157]: talked IRL

File: packages/zendesk-adapter/src/filters/article/article_body.ts [hunk 8]
```diff
@@ -222,30 +152,25 @@ const getWarningsForMissingBrands = (
...
+  )
+  const brandsByUrl = _.keyBy(
+    instances.filter(instance => instance.elemID.typeName === BRAND_TYPE_NAME),
+    i => _.toString(i.value.brand_url)
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/article/article_body.ts:161]: talked IRL

File: packages/zendesk-adapter/src/filters/article/article_body.ts (hunk 2 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/article/article_body.ts:102]: My first thought was that in article body there are links specifically to articles and items inside the brand, so it makes sense to search for brand there
it is possible to expand the search for brand to everything, but i need to think what to do about reference expression

possible solution is to add '.zendesk' to the brand regex, i'll give it a thought and some research and update

File: packages/zendesk-adapter/src/filters/utils.ts [hunk 9]
```diff
@@ -146,3 +160,67 @@ export const getCustomFieldOptionsFromChanges = (parentTypeName: string, childTy
...
+  urlPart: string
+  urlBrandInstance?: InstanceElement
+  instancesById: Record<string, InstanceElement>
+  idRegex: RegExp
+  type: string
+  enableMissingReferences?: boolean
+}): TemplatePart[] | undefined => {
+  const { url, id } = urlPart.match(idRegex)?.groups ?? {}
+  if (url !== undefined && id !== undefined) {
+    const referencedInstance = instancesById[id]
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/utils.ts:191]: The check about the brandId was to make sure that the categories/article... is of the brand that shows in the subdomain

because now we know that id is unique across all brands, this check doesn't seem to do anything but complicating the code

File: packages/zendesk-adapter/test/filters/article/article_body.test.ts [hunk 10]
```diff
@@ -63,125 +73,144 @@ describe('article body filter', () => {
     },
   })
 
-  const brandInstance = new InstanceElement(
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/test/filters/article/article_body.test.ts:66]: The only test i removed was the one that tested handling of an element not being a child of the brand shown in the subdomain (because i also removed that check in the code)

the only other change is making the ids unique

File: packages/zendesk-adapter/src/filters/utils.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (netama) [reply packages/zendesk-adapter/src/filters/utils.ts:190]: but even if it's unique it might be misconfigured, no? if you look for an id under the wrong brand the url won't work

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts [hunk 11]
```diff
@@ -342,11 +396,19 @@ export const prepRef = (part: ReferenceExpression): TemplatePart => {
   if (part.elemID.typeName === GROUP_TYPE_NAME && part.value?.value?.id) {
     return part.value.value.id.toString()
   }
+  if (extractReferencesFromFreeText
+    && ELEMENTS_REGEXES.map(({ type }) => type).includes(part.elemID.typeName) && part.value?.value?.id) {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/handle_template_expressions.ts:400]: let's turn the types from `ELEMENTS_REGEXES` into a set? this calculation is run a lot

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts [hunk 12]
```diff
@@ -322,7 +376,7 @@ const replaceFormulasWithTemplates = (instances: InstanceElement[], enableMissin
   }
 }
 
-export const prepRef = (part: ReferenceExpression): TemplatePart => {
+export const prepRef = (part: ReferenceExpression, extractReferencesFromFreeText?: boolean): TemplatePart => {
```
Comment (netama) [inline packages/zendesk-adapter/src/filters/handle_template_expressions.ts:379]: this function is called from other places as well (e.g. `organization_field`). shouldn't they also get the new behavior? seems they now depend on passing this extra parameter.

**Review by netama — APPROVED** [review]
added a few more comments, in general this looks good so pre-approving to unblock, but if there are big changes that are different from the original code I will do another pass. thanks

File: packages/zendesk-adapter/src/filters/utils.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/utils.ts:190]: it might be if someone configured it manually, which i don't see happening because if the subdomain is included, it will be either a built in link (which cant be wrong) or a link the user put (which would probably be copy-paste)

keeping it may keep the user safe from mistyping it manual, but i think the complication of the code is not worth it (i wrote it and it took me a minute to understand what was going on)

File: packages/zendesk-adapter/src/filters/handle_template_expressions.ts (hunk 12 above)
Comment (idoser96) [reply packages/zendesk-adapter/src/filters/handle_template_expressions.ts:379]: this is only called from organization_field as well, and there is passes raw_name, which is not really a free text (its a title that can be dynamic_content)

because not passing the parameter keeps the old behaviour, it doesn't change anything there

File: packages/zendesk-adapter/e2e_test/adapter.test.ts [hunk 13]
```diff
@@ -848,7 +848,7 @@ describe('Zendesk adapter E2E', () => {
           body: new TemplateExpression({
             parts: [
               '<p>this is a test <img src="',
-              new ReferenceExpression(brandInstanceE2eHelpCenter.elemID.createNestedID('brand_url'), brandInstanceE2eHelpCenter?.value.brand_url),
+              new ReferenceExpression(brandInstanceE2eHelpCenter.elemID, brandInstanceE2eHelpCenter),
```
Comment (edenhassid) [inline packages/zendesk-adapter/e2e_test/adapter.test.ts:851]: why did you remove this? why not check both?

File: packages/zendesk-adapter/src/filters/article/article_body.ts [hunk 14]
```diff
@@ -122,7 +142,9 @@ const updateArticleTranslationBody = ({
  */
 export const prepRef = (part: ReferenceExpression): TemplatePart => {
   if (part.elemID.typeName === BRAND_TYPE_NAME) {
-    return part.value.value.brand_url
+    // The value used to be a string, but now it's an instance element
```
Comment (edenhassid) [inline packages/zendesk-adapter/src/filters/article/article_body.ts:145]: why its an instance element now? I don't understand

**Review by edenhassid — COMMENTED** [review]
looks good, added a few comments

File: packages/zendesk-adapter/src/config.ts [hunk 15]
```diff
@@ -84,6 +84,7 @@ export type ZendeskFetchConfig = configUtils.UserFetchConfig
   appReferenceLocators?: IdLocator[]
   guide?: Guide
   resolveOrganizationIDs?: boolean
+  transformLinks?: boolean
```
[the remark this answers is not in the document]
Comment (idoser96) [reply packages/zendesk-adapter/src/config.ts]: the default is false so regular users won't have sudden change, currently this is a feature for a specific customer
